libs/agno/agno/agent/agent.py

```diff
@@ -1235,106 +1312,1080 @@ async def arun(
-                self.run_response = run_response
-                self.run_id = run_id
+        if stream_intermediate_steps:
+            yield self.create_run_response(
+                content=run_response.content,
+                reasoning_content=run_response.reasoning_content,
+                session_id=session_id,
+                event=RunEvent.run_completed,
+                run_response=run_response,
+            )
 
-                # Set run_input
-                if message is not None:
-                    if isinstance(message, str):
-                        self.run_input = message
-                    elif isinstance(message, Message):
-                        self.run_input = message.to_dict()
-                    else:
-                        self.run_input = message
-                elif messages is not None:
-                    self.run_input = [m.to_dict() if isinstance(m, Message) else m for m in messages]
+    def _handle_tool_call_updates(self, run_response: RunResponse, run_messages: RunMessages):
+        self.model = cast(Model, self.model)
+        for _t in run_response.tools or []:
+            # Case 1: Handle confirmed tools and execute them
+            if _t.requires_confirmation is not None and _t.requires_confirmation is True and self._functions_for_model:
+                for func_name, func in self._functions_for_model.items():
+                    if func_name == _t.tool_name:
+                        func.requires_confirmation = False
+                # Tool is confirmed and hasn't been run before
+                if _t.confirmed is not None and _t.confirmed is True and _t.result is None:
+                    # Execute the tool
+                    function_call = self.model.get_function_call_to_run_from_tool_execution(
+                        _t, self._functions_for_model
+                    )
+                    function_call_results: List[Message] = []
 
-                log_debug(f"Agent Run Start: {run_response.run_id}", center=True)
+                    for call_result in self.model.run_function_call(
+                        function_call=function_call,
+                        function_call_results=function_call_results,
+                    ):
+                        if (
+                            call_result.event == ModelResponseEvent.tool_call_completed.value
+                            and call_result.tool_executions
+                        ):
+                            tool_execution = call_result.tool_executions[0]
+                            _t.result = tool_execution.result
+                            _t.tool_call_error = tool_execution.tool_call_error
+                    if len(function_call_results) > 0:
+                        run_messages.messages.extend(function_call_results)
+                    _t.requires_confirmation = False
+                else:
+                    raise ValueError(f"Tool {_t.tool_name} requires confirmation, cannot continue run")
```

Comment:
Should we raise or just exit gracefully? I feel it's not really an error

Reply:
Makes sense. Better DX for later would be to prompt the model again for it to find a different solution (instead of calling the tool I mean). It's the conventional behavior in agentic coding tools when you reject a confirmation I think